lib/vnet/setup.go

```diff
@@ -77,7 +101,14 @@ func Run(ctx context.Context, appProvider AppProvider) error {
 		select {
 		case adminCommandErr = <-adminCommandErrCh:
 			// The admin command exited before the context was canceled, cancel everything and exit.
-			cancel()
+			// This can happen if the admin subcommand crashes or gets killed.
+
+			// If socket gets removed, the admin subcommand assumes it's because the process on the other
+			// end of this socket (running this code) has quit.
+			if adminCommandErr == nil {
+				adminCommandErr = trace.Errorf("admin subcommand exited prematurely with no error (likely because socket was removed)")
```

Comment:
note: i think i need to improve/change the socket based keepalive for the admin process, i noticed it never exits if the non-root process gets killed without getting a chance to close/delete the socket, like when i run with a debugger. probably need to go back to the PID approach

Reply:
The socket approach was so nice from Connect's perspective though, because I could just shut down the socket and the admin process would shut down too. With the PID approach, the only thing that comes to my mind is hacks like removing the PID file to make the admin process think that the unprivileged process has quit.

Reply:
Oh right, yeah let's avoid any PID approach. Another idea is to keep the user process listening on the socket and have the root process make sure it can continually dial to it, that way if the user process dies without deleting the socket, the dial will fail, and the root process can clean up

Reply:
Gah that would be bad when running with a debugger too though, if the user process doesn't get a chance to accept a connection from the socket often enough

Reply:
Did you already experiment with passing an extra file descriptor to osascript and then weaving it through to the admin subcommand? https://stackoverflow.com/a/18732806/742872

This way we could communicate through a pipe with the admin subcommand. OTOH the example above doesn't use "with administrator privileges", I guess osascript probably doesn't forward fds when it does that.

Reply:
Ultimately I think we should do a PID file at a static path `/var/run/vnet.pid`. This can prevent a user from running multiple VNet processes at once, and also solve the admin process termination issue

We should only create it when starting VNet and delete it when stopping, Connect should be able to handle that, it won't necessarily be tied to the full process lifetime

When tsh or Connect try to start VNet, they should check if this file exists. If not, they should create it with their own PID.

if the file does exist, check if the process with the PID is actually running. If not, assume it died without cleaning up the file, and overwrite it. If that process is running, return an error and fail to start vnet since it is already running in another process

the admin process should watch that file and terminate when either: it is deleted or the PID in it changes or the process associated with the PID in the file is not running.

I'm not sure where to make this change at this point to reduce conflicts with you. Maybe after you get these current lifecycle improvement PRs sorted out, I can base it off your newest branch